Add ERC: Store, Table-Based Introspectable Storage (#711)
ethereum/ERCs: ERCS/erc-7813.md

---
eip: 7813
title: Store, Table-Based Introspectable Storage
description: On-chain tables for automatic indexing and introspectable state
author: alvarius (@alvrs), dk1a (@dk1a), frolic (@holic), ludens (@ludns), vdrg (@vdrg), yonada <yonada@proton.me>
discussions-to: https://ethereum-magicians.org/t/erc-7813-store-table-based-introspectable-storage/21628
status: Draft
type: Standards Track
category: ERC
created: 2024-11-08
---

## Abstract

> This standard introduces a flexible on-chain storage pattern that organizes data into structured tables with schemas, similar to a traditional database. This approach allows new tables to be added at runtime without impacting existing contracts, thereby simplifying upgrades and extensions. By providing a unified interface for data access, the standard enables any contract or off-chain service to read stored data without the need for custom getter functions. Additionally, by standardizing event emissions for state changes it enables automatic, schema-aware indexing.

Reviewer: Your abstract is missing some technical meat. After reading this section, your audience should have a high level (but still technical) overview of _how_ your proposal accomplishes its goals.

3. **Simpler Upgradability**: This pattern leverages unstructured storage, making it easier to upgrade contract logic without the risks associated with using a fixed storage layout.
4. **Flexible Data Extensions**: New tables can be added at runtime without without breaking existing integrations with other data consumers.
5. **Reduced gas costs**: Using efficient data packing reduces gas costs for both storage and event emissions.

## Specification

### Definitions

#### Store

A smart contract that implements the interface proposed by this ERC and organizes data in Tables. It emits events for each data operation so that off-chain components can replicate the state of all tables.

#### Table

A storage structure that holds **Records** sharing the same **Schema**.

- **On-chain Table**: Stores its state on-chain and emits events for off-chain- indexers.
- **Off-chain Table**: Does not store state on-chain but emits events for off-chain indexers.

#### Record

A piece of data stored in a **Table**, addressed by one or more keys.

> #### ResourceId

Reviewer: If these headings are literals from your code (and it looks like they might be), you should enclose them in backticks (\`). That goes for any code snippets in the text as well.

   */
  function getFieldLength(
    ResourceId tableId,
    bytes32[] memory keyTuple,
    uint8 fieldIndex
  ) external view returns (uint256);
}
```

The return values of both `getRecord` and `getField` use the encoding rules previously defined in the Packed Data Encoding section. More specifically, `getRecord` returns the fully encoded record tuple, and the data returned by `getField` is encoded using the encoding rules as if the field was being encoded on its own.

### Store Operations and Events

This standard defines three core operations for manipulating records in a table: setting, updating, and deleting. For each operation, specific events must be emitted. The implementation details of these operations are left to the discretion of each Store implementation.

The fundamental requirement is that for on-chain tables the Record data retrieved through the Store interface methods at any given block MUST be consistent with the Record data that would be obtained by applying the operations implied by the Store events up to that block. This ensures data integrity and allows for accurate off-chain state reconstruction.

#### Store_SetRecord

> Setting a Record means overwriting all of its fields. This operation can be performed wether the record has been set before or not (the standard does not enforce existence checks).

Reviewer: ```suggestion
Setting a Record means overwriting all of its fields. This operation can be performed whether the record has been set before or not (the standard does not enforce existence checks).
```

> The `Store_SetRecord` event **MUST** be emitted to signal that the full data of a record has been overwritten.

Reviewer: ```suggestion
The `Store_SetRecord` event **MUST** be emitted whenever the full data of a record has been overwritten.
```

Is this more clear?

  bytes dynamicData
);
```

Parameters:

| **Name**       | **Type**       | **Description**                                                                              |
| -------------- | -------------- | -------------------------------------------------------------------------------------------- |
| tableId        | ResourceId     | The ID of the table where the record is set                                                  |
| keyTuple       | bytes32[]      | An array representing the composite key for the record                                       |
| staticData     | bytes          | The static data of the record using packed encoding                                          |
| encodedLengths | EncodedLengths | The encoded lengths of the dynamic data of the record                                        |
| dynamicData    | bytes          | The dynamic data of the record, using custom packed encoding (TODO link to encoding section) |

#### Store_SpliceStaticData

Splicing the static data of a Record consists in overwriting bytes of the packed encoded static fields. The total length of static data does not change as it is determined by the table’s value schema.

> The `Store_SpliceStaticData` event MUST be emitted to signal that the static data of the Record has been spliced.

Reviewer: ```suggestion
The `Store_SpliceStaticData` event MUST be emitted whenever the static data of the Record has been spliced.
```

```solidity
event Store_SpliceStaticData(
  ResourceId indexed tableId,
  bytes32[] keyTuple,
  uint48 start,
  bytes data
);
```

Parameters:

| **Name** | **Type**   | **Description**                                               |
| -------- | ---------- | ------------------------------------------------------------- |
| tableId  | ResourceId | The ID of the table where the data is spliced                 |
| keyTuple | bytes32[]  | An array representing the key for the record                  |
| start    | uint48     | The start position in bytes for the splice operation          |
| data     | bytes      | Packed ABI encoding of a tuple with the value's static fields |

#### Store_SpliceDynamicData

Splicing the dynamic data of a Record involves modifying the packed encoded representation of its dynamic fields by removing, replacing, and/or inserting new bytes in place.

> The `Store_SpliceDynamicData` event MUST be emitted to signal that the dynamic data of the Record has been spliced.

Reviewer: ```suggestion
The `Store_SpliceDynamicData` event MUST be emitted whenever the dynamic data of the Record has been spliced.
```

| keyTuple          | bytes32[]      | An array representing the composite key for the record                                                                                                   |
| dynamicFieldIndex | uint8          | The index of the dynamic field to splice data, relative to the start of the dynamic fields (Dynamic field index = field index - number of static fields) |
| start             | uint48         | The start position in bytes for the splice operation                                                                                                     |
| deleteCount       | uint40         | The number of bytes to delete in the splice operation                                                                                                    |
| encodedLengths    | EncodedLengths | The resulting encoded lengths of the dynamic data of the record                                                                                          |
| data              | bytes          | The data to insert into the dynamic data of the record at the start byte                                                                                 |

#### Store_DeleteRecord

> The `Store_DeleteRecord` event MUST be emitted to signal that the Record has been deleted from the Table.

Reviewer: ```suggestion
The `Store_DeleteRecord` event MUST be emitted whenever the Record has been deleted from the Table.
```

```solidity
event Store_DeleteRecord(ResourceId indexed tableId, bytes32[] keyTuple);
```

Parameters:

| **Name** | **Type**   | **Description**                                        |
| -------- | ---------- | ------------------------------------------------------ |
| tableId  | ResourceId | The ID of the table where the record is deleted        |
| keyTuple | bytes32[]  | An array representing the composite key for the record |

> #### Example

Reviewer: This should probably be moved to your Reference Implementation section. You can link to it from here though.

```solidity
event Store_SetRecord(
  ResourceId indexed tableId,
  bytes32[] keyTuple,
  bytes staticData,
  EncodedLengths encodedLengths,
  bytes dynamicData
);
```

Parameters:

| **Name**       | **Type**       | **Description**                                                                              |
| -------------- | -------------- | -------------------------------------------------------------------------------------------- |
| tableId        | ResourceId     | The ID of the table where the record is set                                                  |
| keyTuple       | bytes32[]      | An array representing the composite key for the record                                       |
| staticData     | bytes          | The static data of the record using packed encoding                                          |
| encodedLengths | EncodedLengths | The encoded lengths of the dynamic data of the record                                        |
| dynamicData    | bytes          | The dynamic data of the record, using [custom packed encoding](#packed-data-encoding)        |

> #### Store_SpliceStaticData

Reviewer: ```suggestion
#### `Store_SpliceStaticData`
```

Splicing the static data of a Record consists in overwriting bytes of the packed encoded static fields. The total length of static data does not change as it is determined by the table’s value schema.

The `Store_SpliceStaticData` event MUST be emitted whenever the static data of the Record has been spliced.

```solidity
event Store_SpliceStaticData(
  ResourceId indexed tableId,
  bytes32[] keyTuple,
  uint48 start,
  bytes data
);
```

Parameters:

| **Name** | **Type**   | **Description**                                               |
| -------- | ---------- | ------------------------------------------------------------- |
| tableId  | ResourceId | The ID of the table where the data is spliced                 |
| keyTuple | bytes32[]  | An array representing the key for the record                  |
| start    | uint48     | The start position in bytes for the splice operation          |
| data     | bytes      | Packed ABI encoding of a tuple with the value's static fields |

> #### Store_SpliceDynamicData

Reviewer: ```suggestion
#### `Store_SpliceDynamicData`
```

| keyTuple          | bytes32[]      | An array representing the composite key for the record                                                                                                   |
| dynamicFieldIndex | uint8          | The index of the dynamic field to splice data, relative to the start of the dynamic fields (Dynamic field index = field index - number of static fields) |
| start             | uint48         | The start position in bytes for the splice operation                                                                                                     |
| deleteCount       | uint40         | The number of bytes to delete in the splice operation                                                                                                    |
| encodedLengths    | EncodedLengths | The resulting encoded lengths of the dynamic data of the record                                                                                          |
| data              | bytes          | The data to insert into the dynamic data of the record at the start byte                                                                                 |

> #### Store_DeleteRecord

Reviewer: ```suggestion
#### `Store_DeleteRecord`
```
